lib/semantics/check-select-stmt.cc

```diff
@@ -0,0 +1,239 @@
+                                 (valPair.second.value() >= mem.first))) &&
+                            (!mem.second ||
+                                (valPair.second &&
+                                    (valPair.second.value() <= mem.second)))));
+                  });
+              if (intIter == caseRangeIntVals.end()) {
+                caseRangeIntVals.push_back(valPair);
+                continue;
+              }
+            }
+          } else {  // TypeCategory::Character
+            const auto &strPair{std::make_pair(lower
+                    ? std::optional(GetString(lower->thing).value())
+                    : std::nullopt,
+                upper ? std::optional(GetString(upper->thing).value())
+                      : std::nullopt)};
+            if ((!lower ||
+                    (caseVals.find(strPair.first.value()) == caseVals.end())) &&
+                (!upper ||
+                    (caseVals.find(strPair.second.value()) ==
+                        caseVals.end()))) {
+              stringIter = std::find_if(caseRangeStringVals.begin(),
+                  caseRangeStringVals.end(), [&strPair](const auto &mem) {
+                    return ((!mem.first && !strPair.first) ||
+                        (!mem.second && !strPair.second) ||
+                        ((!mem.first ||
+                             (strPair.first &&
+                                 ((strPair.first.value())
+                                         .compare(mem.first.value()) > 0))) &&
+                            (!mem.second ||
+                                (strPair.first &&
+                                    ((strPair.first.value())
+                                            .compare(mem.second.value()) <
+                                        0)))) ||
+                        ((!mem.first ||
+                             (strPair.second &&
+                                 (strPair.second.value().compare(
+                                      mem.first.value()) > 0))) &&
+                            (!mem.second ||
+                                (strPair.second &&
+                                    (strPair.second.value().compare(
+                                         mem.second.value()) < 0)))));
+                  });
+
+              if (stringIter == caseRangeStringVals.end()) {
+                caseRangeStringVals.push_back(strPair);
+                continue;
+              }
+            }
+          }
+          // C1149
+          context_.Say(caseStmt.source,
+              "SELECT CASE statement value must not match more than one case-value-range"_err_en_US);
+        }
+      }
+    }
+  }
+}
+
+bool SelectStmtChecker::isValidSelectCaseType(const TypeCategory &expectedType,
+    const parser::ConstantExpr &constCase, const parser::CharBlock &src) {
+  if (const auto *caseval{GetExpr(constCase)}) {
+    if (auto type{caseval->GetType()}) {
+      if (type->category() != expectedType) {  // C1147 (R1140)
```

Comment:
C1147 also says that character kinds must match. Are you checking that?

Reply:
I have handled character kinds as well.